### PR TITLE
Solve memory inefficiency in RNNs

### PR DESCRIPTION
**Pull request by atmguille**
As discussed in [this issue](https://github.com/keras-team/keras/issues/16113) with @mattdangerw, I am opening this PR to solve memory inefficiency in RNNs.

@mattdangerw @qlzh727 @fchollet

### CODE REVIEW

File: keras/backend.py [hunk 1]
```diff
@@ -4430,13 +4431,19 @@ def rnn(step_function,
       zero_output_for_mask: Boolean. If True, the output for masked timestep
           will be zeros, whereas in the False case, output from previous
           timestep is returned.
+      return_all_outputs: Boolean. If True, all outputs of the process will be
```
Comment (fchollet) [inline keras/backend.py:4434]: """
If True, return the recurrent outputs for all timesteps in the sequence.
If False, only return the output for the last timestep (which consumes
less memory).
"""

File: keras/backend.py [hunk 2]
```diff
@@ -4430,13 +4431,19 @@ def rnn(step_function,
...
+          outputs=[last_output] is returned.
 
   Returns:
       A tuple, `(last_output, outputs, new_states)`.
           last_output: the latest output of the rnn, of shape `(samples, ...)`
-          outputs: tensor with shape `(samples, time, ...)` where each
-              entry `outputs[s, t]` is the output of the step function
-              at time `t` for sample `s`.
+          outputs:
+              - If `return_all_outputs`: a tensor with shape
```
Comment (fchollet) [inline keras/backend.py:4443]: `return_all_outputs=True`

File: keras/backend.py [hunk 3]
```diff
@@ -4430,13 +4431,19 @@ def rnn(step_function,
...
+              `(samples, time, ...)` where each entry `outputs[s, t]` is the
+              output of the step function at time `t` for sample `s`
+              - Else, a tensor equal to last_output
```
Comment (fchollet) [inline keras/backend.py:4446]: Use backticks for code keywords

File: keras/backend.py [hunk 4]
```diff
@@ -4597,10 +4612,12 @@ def _get_input_tensor(time):
     # the value is discarded.
     output_time_zero, _ = step_function(
         input_time_zero, tuple(initial_states) + tuple(constants))
+
+    output_ta_size = time_steps_t if return_all_outputs else 1
```
Comment (fchollet) [inline keras/backend.py:4616]: When `return_all_outputs=False` we don't need the TensorArray logic at all, right?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/backend.py (hunk 4 above)
Comment (atmguille) [reply keras/backend.py:4616]: Totally agree @fchollet . I just wanted to change the logic as little as possible. By having a common TensorArray, the only things that change when `return_all_outputs=False` is the size and that we always write at index 0; similar to what I did in the `unroll=True` branch when keeping a list of one element.

Of course, suggestions are more than welcome to improve this :)

**Review by fchollet — APPROVED** [review]
LGTM, thanks

**Review by fchollet — CHANGES_REQUESTED** [review]
There are test failures, please take a look: https://source.cloud.google.com/results/invocations/919aaccc-7dd1-4d0e-9912-0596f18e60c1/targets/keras%2Fgithub%2Fubuntu%2Fgpu%2Fpresubmit/log

**Comment by atmguille** [comment]
@fchollet just saw the test log. I think it comes from the following: 
`LSTM` and `GRU` have different functions to execute in CPU and GPU (the other RNNs layers don't). CPU version calls `backend.rnn` (the function I updated), but GPU calls `tf.raw_ops.CudnnRNN`. I updated the CPU version to properly call the new version of `backend.rnn`, but the GPU one will always return `outputs` as `tf.raw_ops.CudnnRNN` always returns the full array.

The error happens in the tests that combine CPU and GPU environments.

I see three options to solve this:
1. Deactivate new functionality in `LSTM` and `GRU`. As mentioned in the original issue #16113, the layers most affected by the current behavior are convolutional RNNs, for which the update works perfectly.
2. Although `tf.raw_ops.CudnnRNN` returns `outputs`, the GPU method can still return `outputs=[last_output]`. The memory advantage will not work in GPU, but the return format from both functions would be consistent.
3. Update `tf.raw_ops.CudnnRNN` in a similar way than `backend.rnn`. However, I have not been able to find the source code of that function in the [tensorflow repository](https://github.com/tensorflow/tensorflow).

I think the ideal solution would be 3. For the time being, I would go for 2. Let me know what you think!

PS: I had executed the tests locally with no errors. I have re-executed them again and no errors were found, as can be seen in [this notebook](https://www.kaggle.com/atmguille/keras-pr-gpu-tests).

**Comment by fchollet** [comment]
Let's do option 2.  Option 3. is not realistically doable given that the computation is delegated to cuDNN.

**Comment by atmguille** [comment]
I believe it should be fixed now. I have executed all tests locally and there are no errors. However, there were no errors also before this fix, as I mentioned in my last comment. Let's see

Thanks again for your time @fchollet

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Comment by fchollet** [comment]
I think this is possibly a case of flaky test. Let's try rerunning the tests.

**Comment by atmguille** [comment]
The tests have passed! I am not sure if you need to approve my last review request, as the PR is "Approved by Reviewer" but it also shows that there is 1 pending reviewer.

Anyway, thanks a lot for your time! It was a good learning experience as my first contribution @fchollet
